### PR TITLE
COMMON: reduce mass add detection output by skipping certain ADGF flags and skipping incomplete file/md5/size matches

### PR DESCRIPTION
**Pull request by athrxx**
Skip over targets with ADGF_WARNING and ADGF_UNSUPPORTED flags when using "mass add".

The alternative would/could be to just invent a new flag like ADGF_NOMASSADD

For me, this is mostly about fixing bug no. 13282.
https://bugs.scummvm.org/ticket/13282

We sometimes have bogus entries which only have the purpose of presenting the error message (reasons for being unsupported) contained in the extra field of the detection entry.

This PR certainly does not attend to all issues that the "mass add" option currently has, but it does clean it up a bit...

### CODE REVIEW

**Comment by athrxx** [comment]
I have added another commit that also skips over all results with incomplete file/md5/size matches in mass add mode (basically treating all files as if they're on the AD grayList). I have tested it on the Kyra engine. It now stops detection Donald Duck's Playground and Glassy Ocean in the Eye of the Beholder directories (although I would still like to add the "start.exe" and the "item.dat" to the grayList, since my current changes only concern the mass add mode).

**Comment by bluegr** [comment]
Nice idea! Why do we want to add engines to toggle these flags? The mass add behavior should be the same across all engines  unless I'm missing something?

**Comment by athrxx** [comment]
Yes, I have used that only  for mass detection.

With that other commit I added, maybe it would be better to just have one extra argument like 'bool massDetectionMode' for the detection method...

**Comment by bluegr** [comment]
What I'm saying is: should we expose these as flags to engines as togglable features, or keep the behavior the same across all engines?

**Comment by athrxx** [comment]
The flags are set in the detection_tables of the engines for each individual target. They are part of the detection entries. Other than that, I am not sure what kind of exposure you mean.

The behavior for all engines: I have really only touched the AdvancedDetector. The few engines that do their own thing (you can easily see from my commits which these are) just ignore the new arguments (except if they also use the AD for some parts of the detection process). Maybe it would make sense to follow up with some handling of the arguments for the rest of the individual detection processes? I don't know. The engine authors who are familiar with their detection code can do that if they want, I guess ..

**Comment by bluegr** [comment]
My bad, these are all overridden methods in the engines. My initial though was to put all this logic inside detectGame() without exposing any extra parameter, but then you can't distinguish between a call to add a single game entry from mass add. Sorry for the confusion :)

**Comment by athrxx** [comment]
Okay, that's good that this is cleared up :-)

So, while we're making progress here, when I now run mass add in my Kyra directory, it still reports "35 new games, ignored 45 previously added games". So we still have 35 too many ;-)
Maybe I can find and fix that bug, too...
I can even limit it to certain game directories. When I run it in my EOB SegaCD English dir it will return "0 new games, ignored 1 previously added games". But when I use my EOB SegaCD Japanese dir it will return "1 new games, ignored 0 previously added games"... Maybe it's just something with non-English targets. But that's just a guess at this point.

**Comment by athrxx** [comment]
I have recently made a fix (committed directly) which further reduces the finds in my Kyra directory.
My impression is that the ones that remain are all valid...

File: base/plugins.cpp [hunk 1]
```diff
@@ -717,7 +717,7 @@ QualifiedGameList EngineManager::findGameInLoadedPlugins(const Common::String &g
 	return results;
 }
 
-DetectionResults EngineManager::detectGames(const Common::FSList &fslist) {
+DetectionResults EngineManager::detectGames(const Common::FSList &fslist, int skipADFlags, bool skipIncomplete) {
```
Comment (sev-) [inline base/plugins.cpp:720]: A major thing: `g->flags` is `uint32`, thus, your `skipADFlags` must be of this type too.

File: engines/advancedDetector.cpp [hunk 2]
```diff
@@ -589,7 +589,7 @@ ADDetectedGames AdvancedMetaEngineDetection::detectGame(const Common::FSNode &pa
 	// they are present. Compute MD5s and file sizes for the available files.
 	for (descPtr = _gameDescriptors; ((const ADGameDescription *)descPtr)->gameId != nullptr; descPtr += _descItemSize) {
 		g = (const ADGameDescription *)descPtr;
-
+
```
Comment (sev-) [inline engines/advancedDetector.cpp:592]: This slipped it.

File: engines/advancedDetector.cpp [hunk 3]
```diff
@@ -627,6 +627,11 @@ ADDetectedGames AdvancedMetaEngineDetection::detectGame(const Common::FSNode &pa
 		if ((_flags & kADFlagUseExtraAsHint) && !extra.empty() && g->extra != extra)
 			continue;
 
+		if (g->flags & skipADFlags) {
+			debugC(3, kDebugGlobalDetection,"Skipping unsupported target for engine '%s' for the following reasons:\n\t'%s'", g->gameId, g->extra);
```
Comment (sev-) [inline engines/advancedDetector.cpp:631]: add space after second comma?

File: engines/advancedDetector.h [hunk 4]
```diff
@@ -419,15 +419,17 @@ class AdvancedMetaEngineDetection : public MetaEngineDetection {
...
-	 * @param allFiles  List of all present files, as computed by the @ref composeFileHashMap.
-	 * @param language  Restrict results to the specified language.
-	 * @param platform  Restrict results to the specified platform.
-	 * @param extra     Restrict results to the specified @c extra string (only if @ref kADFlagUseExtraAsHint is set).
+	 * @param parent		  Parent node of this file node.
+	 * @param allFiles		  List of all present files, as computed by the @ref composeFileHashMap.
+	 * @param language		  Restrict results to the specified language.
+	 * @param platform		  Restrict results to the specified platform.
+	 * @param extra			  Restrict results to the specified @c extra string (only if @ref kADFlagUseExtraAsHint is set).
+	 * @param skipADFlags	  Ignore results with these ADGF flags (for mass add).
```
Comment (sev-) [inline engines/advancedDetector.h:427]: Specify bitmask of ADGF flags to be ignored (for mass add)

**Review by sev- — CHANGES_REQUESTED** [review]
Nice feature, though there are few notes.

**Comment by sev-** [comment]
Excelletn! Merging

**Pull request merged**